lib/src/lis/io.cpp

```diff
@@ -0,0 +1,524 @@
+        if ( prh.attributes & lis::prheader::filenum ) trlen += 2;
+        if ( prh.attributes & lis::prheader::chcksum ) trlen += 2;
+
+        /* LRs spanning multiple PRs: The LRH is only present in the first PRH,
+         * i.e. when the predecessor attribute is not set. The LRH is already
+         * read, and is a part of record_info, so just skip past it.
+         */
+        std::int64_t toread = prh.length - lis::prheader::size - trlen;
+        if ( not (prh.attributes & lis::prheader::predces) ) {
+            this->seek(this->ltell() + lis::lrheader::size);
+            toread -= lis::lrheader::size;
+        }
+
+        rec.data.resize(rec.data.size() + toread);
+
+        auto nread = this->read(rec.data.data() + prevlen, toread);
+
+        if (nread < toread) {
+            throw dlisio::io_error("iodevice::read_record: record truncated");
+        }
+
+        /* Skip past trailer(s) */
+        if (trlen) this->seek(this->ltell() + trlen);
+
+        prevlen += toread;
+
+        if ( not (prh.attributes & lis::prheader::succses) ) break;
+    }
+    return rec;
+}
+
+
+/* miscellaneous */
+iodevice open( const std::string& path, std::int64_t offset, bool tapeimage )
+noexcept (false) {
+    auto* file = std::fopen(path.c_str(), "rb");
+    if ( not file ) {
+        auto msg = "lis::open: unable to open file for path {} : {}";
+        throw dlisio::io_error(fmt::format(msg, path, strerror(errno)));
+    }
+
+    auto* protocol = lfp_cfile(file);
+    if ( protocol == nullptr ) {
+        throw dlisio::io_error("lis::open: "
+                               "lfp: unable to open lfp protocol cfile");
+    }
+
+    auto err = lfp_seek(protocol, offset);
+    switch (err) {
+        case LFP_OK: break;
+        default: {
+            lfp_close( protocol );
+            throw dlisio::io_error( lfp_errormsg(protocol) );
+        }
+    }
+
+    if ( tapeimage ) {
+        auto* tif = lfp_tapeimage_open(protocol);
+        if ( tif == nullptr ) {
+            lfp_close( protocol );
+            throw dlisio::io_error("lis::open: "
+                                   "unable to open lfp protocol tapeimage");
+        }
+        protocol = tif;
+    }
+
+    auto device = iodevice( protocol );
+
+    /* Verify that the device is not opened at EOF by attempting to read one byte */
+    try {
+        char tmp;
+        device.read(&tmp, 1);
+    } catch ( const std::exception& e ) {
+        const auto poffset = device.poffset();
+        device.close();
+        const auto msg = "lis::open: "
+                            "Cannot open lis::iodevice (ptell={}): {}";
+        throw dlisio::io_error( fmt::format(msg, poffset, e.what() ));
+    }
+
+    if ( device.eof() ) {
+        const auto poffset = device.poffset();
+        device.close();
+        const auto msg = "open: handle is opened at EOF (ptell={})";
+        throw dlisio::eof_error( fmt::format(msg, poffset) );
+    }
+
+    try {
+        device.seek( 0 );
```

Comment:
TODO: untiffed file with >1 logical files which reaches here.
Suspect infinite loop due to similar load issue as in dlisio caused by lfp bug.

(Now might not be a problem, seems like index_records won't ever return until reached physical EOF in untiffed, but if we decide to break into logical files here, it would be, I believe)

[tiffed/untiffed: 0/device.pzero, same inconsistency logical issue as in dlis.load due to lfp bug]